crates/ethereum/payload/src/lib.rs

```diff
@@ -202,32 +179,24 @@ where
     let block_number = evm_env.block_env.number;
     let beneficiary = evm_env.block_env.beneficiary;
 
-    let mut evm = evm_config.evm_with_env(&mut db, evm_env);
-    let mut system_caller = SystemCaller::new(chain_spec.clone());
+    let mut strategy = EthExecutionStrategy::new(
+        evm_config.evm_with_env(&mut db, evm_env),
+        EthBlockExecutionInput {
+            number: parent_header.number + 1,
+            timestamp: attributes.timestamp(),
+            parent_hash: parent_header.hash(),
+            gas_limit: next_attributes.gas_limit,
+            parent_beacon_block_root: attributes.parent_beacon_block_root,
+            beneficiary,
+            ommers: &[],
+            withdrawals: Some(&attributes.withdrawals),
+        },
+        &chain_spec,
+        &evm_config,
+    );
+
+    strategy.apply_pre_execution_changes().map_err(PayloadBuilderError::other)?;
 
-    // apply eip-4788 pre block contract call
-    system_caller
-        .apply_beacon_root_contract_call(attributes.parent_beacon_block_root, &mut evm)
-        .map_err(|err| {
-            warn!(target: "payload_builder",
-                parent_hash=%parent_header.hash(),
-                %err,
-                "failed to apply beacon root contract call for payload"
-            );
-            PayloadBuilderError::Internal(err.into())
-        })?;
-
-    // apply eip-2935 blockhashes update
-    system_caller.apply_blockhashes_contract_call(
-        parent_header.hash(),
-        &mut evm,
-    )
-    .map_err(|err| {
-        warn!(target: "payload_builder", parent_hash=%parent_header.hash(), %err, "failed to update parent header blockhashes for payload");
-        PayloadBuilderError::Internal(err.into())
-    })?;
-
-    let mut receipts = Vec::new();
     let mut block_blob_count = 0;
```

Comment:
I would do all this before taking mutable ref to db, do as much as possible before and after to hold exclusive access as short as possible

Reply:
possibly out of scope since the original code wasn't taking this into account either

Reply:
this doesn't really matter because the db is only dropped afterwards

Reply:
it could be dropped earlier though it seems? anyhow, out of scope